File: server/apps/sync-service/src/main/java/com/sharetimer/syncservice/application/service/TimerService.java
package com.sharetimer.syncservice.application.service;

import java.time.Instant;
import org.springframework.http.codec.ServerSentEvent;
import org.springframework.stereotype.Service;
import com.sharetimer.storage.redis.config.ReactiveRedisTemplateFactory;
import com.sharetimer.storage.redis.config.TimerRedisProps;
import com.sharetimer.syncservice.adapter.out.external.TimerApiClient;
import com.sharetimer.syncservice.application.port.in.TimerUseCase;
import com.sharetimer.syncservice.application.port.out.TimerEventPublisher;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import reactor.core.publisher.Flux;

@Service
@RequiredArgsConstructor
@Slf4j
public class TimerService implements TimerUseCase {

  private final TimerEventPublisher timerEventPublisher;
  private final ReactiveRedisTemplateFactory redisTemplateFactory;
  private final TimerRedisProps timerRedisProps;
  private final TimerApiClient timerApiClient;

  @Override
  public Flux<ServerSentEvent<Object>> subscribe(String timerId) {
    String key = timerRedisProps.getExpiration().getKeyPrefix() + timerId;
    return redisTemplateFactory.getTemplate(timerRedisProps.getExpiration().getDbIndex())
        .hasKey(key).flatMapMany(hasKey -> {
          if (!Boolean.TRUE.equals(hasKey)) {
            return Flux.just(ServerSentEvent.builder().event("timerNotFound").build());
          }
          return timerEventPublisher.subscribe(timerId);
        });
  }

  @Override
  public void processTimerExpiration(String timerId) {
    timerEventPublisher.sendTimerEndEvent(timerId);
    timerApiClient.deleteTimer(timerId).subscribe(
        success -> log.debug("Timer delete request success: {}", timerId),
        error -> log.error("Timer delete request failed: {}", timerId, error));
  }

  @Override
  public void updateTargetTime(String timerId, Instant updatedAt, Instant serverTime,
      Instant newTargetTime) {
    timerEventPublisher.updateTargetTime(timerId, updatedAt, serverTime, newTargetTime);
  }

  @Override
  public void addTimestamp(String timerId, Instant targetTime, Instant capturedAt) {
    timerEventPublisher.addTimestamp(timerId, targetTime, capturedAt);
  }
}

File: server/apps/sync-service/src/main/kotlin/com/sharetimer/syncservice/application/service/TimerService.kt
package com.sharetimer.syncservice.application.service

import com.sharetimer.storage.redis.config.ReactiveRedisTemplateFactory
import com.sharetimer.storage.redis.config.TimerRedisProps
import com.sharetimer.syncservice.adapter.output.external.TimerApiClient
import com.sharetimer.syncservice.application.port.input.TimerUseCase
import com.sharetimer.syncservice.application.port.output.TimerEventPublisher
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.reactor.awaitSingle
import org.slf4j.LoggerFactory
import org.springframework.http.codec.ServerSentEvent
import org.springframework.stereotype.Service
import java.time.Instant

@Service
class TimerService(
    private val timerEventPublisher: TimerEventPublisher,
    private val redisTemplateFactory: ReactiveRedisTemplateFactory,
    private val timerRedisProps: TimerRedisProps,
    private val timerApiClient: TimerApiClient,
) : TimerUseCase {
    override fun subscribe(timerId: String): Flow<ServerSentEvent<Any>> {
        val key = timerRedisProps.expiration.keyPrefix + timerId
        return flow {
            val hasKey = redisTemplateFactory.template.hasKey(key).awaitSingle()
            if (!hasKey) {
                emit(ServerSentEvent.builder<Any>().event("timerNotFound").build())
            } else {
                emitAll(timerEventPublisher.subscribe(timerId))
            }
        }
    }

    override fun processTimerExpiration(timerId: String) {
        timerEventPublisher.sendTimerEndEvent(timerId)
        timerApiClient.deleteTimer(timerId).subscribe(
            { log.debug("Timer delete request success: {}", timerId) },
            { error -> log.error("Timer delete request failed: {}", timerId, error) },
        )
    }

    override fun updateTargetTime(
        timerId: String,
        updatedAt: Instant,
        serverTime: Instant,
        newTargetTime: Instant,
    ) {
        timerEventPublisher.updateTargetTime(timerId, updatedAt, serverTime, newTargetTime)
    }

    override fun addTimestamp(
        timerId: String,
        targetTime: Instant,
        capturedAt: Instant,
    ) {
        timerEventPublisher.addTimestamp(timerId, targetTime, capturedAt)
    }

    companion object {
        private val log = LoggerFactory.getLogger(TimerService::class.java)
    }
}
